File: sketch/src/main/java/me/panpf/sketch/viewfun/ShowDownloadProgressFunction.java
/*
 * Copyright (C) 2013 Peng fei Pan <sky@panpf.me>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.panpf.sketch.viewfun;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.view.View;

import me.panpf.sketch.SLog;
import me.panpf.sketch.decode.ImageAttrs;
import me.panpf.sketch.request.CancelCause;
import me.panpf.sketch.request.DisplayCache;
import me.panpf.sketch.request.ErrorCause;
import me.panpf.sketch.request.ImageFrom;
import me.panpf.sketch.shaper.ImageShaper;
import me.panpf.sketch.uri.UriModel;

/**
 * 显示下载进度功能，会在 {@link android.widget.ImageView} 上面显示一个黑色半透明蒙层显示下载进度，蒙层会随着进度渐渐变小
 */
public class ShowDownloadProgressFunction extends ViewFunction {
    static final int DEFAULT_MASK_COLOR = 0x22000000;

    private static final String NAME = "ShowProgressFunction";
    private static final int NONE = -1;

    private FunctionPropertyView view;
    private int maskColor = DEFAULT_MASK_COLOR;
    private ImageShaper maskShaper;

    private Paint maskPaint;
    private float progress = NONE;
    private Rect bounds;

    public ShowDownloadProgressFunction(@NonNull FunctionPropertyView view) {
        this.view = view;
    }

    @Override
    public boolean onReadyDisplay(@Nullable UriModel uriModel) {
        long newProgress = uriModel != null && uriModel.isFromNet() ? 0 : NONE;
        boolean needRefresh = progress != newProgress;
        progress = newProgress;
        return needRefresh;
    }

    @Override
    public void onDraw(@NonNull Canvas canvas) {
        if (progress == NONE) {
            return;
        }

        ImageShaper shaper = getMaskShaper();
        if (shaper != null) {
            canvas.save();
            try {
                if (bounds == null) {
                    bounds = new Rect();
                }
                bounds.set(view.getPaddingLeft(), view.getPaddingTop(), view.getWidth() - view.getPaddingRight(), view.getHeight() - view.getPaddingBottom());
                Path maskPath = shaper.getPath(bounds);
                canvas.clipPath(maskPath);
            } catch (UnsupportedOperationException e) {
                SLog.e(NAME, "The current environment doesn't support clipPath has shut down automatically hardware acceleration");
                view.setLayerType(View.LAYER_TYPE_SOFTWARE, null);
                e.printStackTrace();
            }
        }

        if (maskPaint == null) {
            maskPaint = new Paint();
            maskPaint.setColor(maskColor);
            maskPaint.setAntiAlias(true);
        }
        canvas.drawRect(
                view.getPaddingLeft(),
                view.getPaddingTop() + (progress * view.getHeight()),
                view.getWidth() - view.getPaddingLeft() - view.getPaddingRight(),
                view.getHeight() - view.getPaddingTop() - view.getPaddingBottom(),
                maskPaint);

        if (shaper != null) {
            canvas.restore();
        }
    }

    @Override
    public boolean onUpdateDownloadProgress(int totalLength, int completedLength) {
        progress = (float) completedLength / totalLength;
        return true;
    }

    @Override
    public boolean onDisplayCompleted(@NonNull Drawable drawable, @NonNull ImageFrom imageFrom, @NonNull ImageAttrs imageAttrs) {
        progress = NONE;
        return true;
    }

    @Override
    public boolean onDisplayError(@NonNull ErrorCause errorCause) {
        progress = NONE;
        return true;
    }

    @Override
    public boolean onDisplayCanceled(@NonNull CancelCause cancelCause) {
        progress = NONE;
        return false;
    }

    public boolean setMaskColor(@ColorInt int maskColor) {
        if (this.maskColor == maskColor) {
            return false;
        }

        this.maskColor = maskColor;
        if (maskPaint != null) {
            maskPaint.setColor(maskColor);
        }
        return true;
    }

    private ImageShaper getMaskShaper() {
        if (maskShaper != null) {
            return maskShaper;
        }

        DisplayCache displayCache = view.getDisplayCache();
        ImageShaper shaperFromCacheOptions = displayCache != null ? displayCache.options.getShaper() : null;
        if (shaperFromCacheOptions != null) {
            return shaperFromCacheOptions;
        }

        ImageShaper shaperFromOptions = view.getOptions().getShaper();
        if (shaperFromOptions != null) {
            return shaperFromOptions;
        }

        return null;
    }

    public boolean setMaskShaper(@Nullable ImageShaper maskShaper) {
        if (this.maskShaper == maskShaper) {
            return false;
        }

        this.maskShaper = maskShaper;
        return true;
    }
}

File: sketch/src/main/java/com/github/panpf/sketch/viewfun/ShowDownloadProgressFunction.kt
/*
 * Copyright (C) 2019 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.panpf.sketch.viewfun

import android.annotation.SuppressLint
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.view.View
import androidx.annotation.ColorInt
import com.github.panpf.sketch.SLog
import com.github.panpf.sketch.decode.ImageAttrs
import com.github.panpf.sketch.request.CancelCause
import com.github.panpf.sketch.request.DisplayCache
import com.github.panpf.sketch.request.ErrorCause
import com.github.panpf.sketch.request.ImageFrom
import com.github.panpf.sketch.shaper.ImageShaper
import com.github.panpf.sketch.uri.UriModel

/**
 * 显示下载进度功能，会在 [android.widget.ImageView] 上面显示一个黑色半透明蒙层显示下载进度，蒙层会随着进度渐渐变小
 */
class ShowDownloadProgressFunction(private val view: FunctionPropertyView) : ViewFunction() {

    private var maskColor = DEFAULT_MASK_COLOR
    private var maskShaper: ImageShaper? = null
    private var maskPaint: Paint? = null
    private var progress = NONE.toFloat()
    private var bounds: Rect? = null

    override fun onReadyDisplay(uri: String): Boolean {
        val uriModel = UriModel.match(view.context, uri)
        val newProgress = if (uriModel != null && uriModel.isFromNet) 0 else NONE.toLong()
        val needRefresh = progress != newProgress.toFloat()
        progress = newProgress.toFloat()
        return needRefresh
    }

    @SuppressLint("DrawAllocation")
    override fun onDraw(canvas: Canvas) {
        if (progress == NONE.toFloat()) {
            return
        }
        val shaper = getMaskShaper()
        if (shaper != null) {
            canvas.save()
            try {
                val bounds = bounds ?: Rect().apply {
                    this@ShowDownloadProgressFunction.bounds = this
                }
                bounds.set(
                    view.paddingLeft,
                    view.paddingTop,
                    view.width - view.paddingRight,
                    view.height - view.paddingBottom
                )
                val maskPath = shaper.getPath(bounds)
                canvas.clipPath(maskPath)
            } catch (e: UnsupportedOperationException) {
                SLog.em(
                    NAME,
                    "The current environment doesn't support clipPath has shut down automatically hardware acceleration"
                )
                view.setLayerType(View.LAYER_TYPE_SOFTWARE, null)
                e.printStackTrace()
            }
        }
        val maskPaint = maskPaint ?: Paint().apply {
            color = maskColor
            isAntiAlias = true
            this@ShowDownloadProgressFunction.maskPaint = this
        }
        canvas.drawRect(
            view.paddingLeft.toFloat(),
            view.paddingTop + progress * view.height,
            (view.width - view.paddingLeft - view.paddingRight).toFloat(),
            (view.height - view.paddingTop - view.paddingBottom).toFloat(),
            maskPaint
        )
        if (shaper != null) {
            canvas.restore()
        }
    }

    override fun onUpdateDownloadProgress(totalLength: Int, completedLength: Int): Boolean {
        progress = completedLength.toFloat() / totalLength
        return true
    }

    override fun onDisplayCompleted(
        drawable: Drawable,
        imageFrom: ImageFrom,
        imageAttrs: ImageAttrs
    ): Boolean {
        progress = NONE.toFloat()
        return true
    }

    override fun onDisplayError(errorCause: ErrorCause): Boolean {
        progress = NONE.toFloat()
        return true
    }

    override fun onDisplayCanceled(cancelCause: CancelCause): Boolean {
        progress = NONE.toFloat()
        return false
    }

    fun setMaskColor(@ColorInt maskColor: Int): Boolean {
        if (this.maskColor == maskColor) {
            return false
        }
        this.maskColor = maskColor
        if (maskPaint != null) {
            maskPaint!!.color = maskColor
        }
        return true
    }

    private fun getMaskShaper(): ImageShaper? {
        val maskShaper = maskShaper
        if (maskShaper != null) {
            return maskShaper
        }
        val displayCache: DisplayCache? = view.displayCache
        val shaperFromCacheOptions = displayCache?.options?.shaper
        return shaperFromCacheOptions ?: view.options.shaper
    }

    fun setMaskShaper(maskShaper: ImageShaper?): Boolean {
        if (this.maskShaper === maskShaper) {
            return false
        }
        this.maskShaper = maskShaper
        return true
    }

    companion object {
        const val DEFAULT_MASK_COLOR = 0x22000000
        private const val NAME = "ShowProgressFunction"
        private const val NONE = -1
    }
}